File: shared/src/main/java/ru/privatenull/pnauth/dialog/DialogAction.java
package ru.privatenull.pnauth.dialog;

import java.util.Map;

/** Every vanilla dialog action, including the two input-aware dynamic variants. */
public sealed interface DialogAction permits DialogAction.None, DialogAction.Static,
        DialogAction.DynamicRunCommand, DialogAction.DynamicCustom {
    record None() implements DialogAction { }
    record Static(String type, Map<String, Object> payload) implements DialogAction {
        public Static { payload = Map.copyOf(payload); }
    }
    record DynamicRunCommand(String template) implements DialogAction { }
    record DynamicCustom(String id, Map<String, Object> additions) implements DialogAction {
        public DynamicCustom { additions = Map.copyOf(additions); }
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/dialog/DialogAction.kt
package ru.privatenull.pnauth.dialog

/** Every vanilla dialog action, including the two input-aware dynamic variants. */
sealed interface DialogAction {
    @JvmRecord
    data class None(val unused: Unit = Unit) : DialogAction {
        constructor() : this(Unit)
    }

    @JvmRecord
    data class Static(val type: String, val payload: Map<String, Any> = emptyMap()) : DialogAction

    @JvmRecord
    data class DynamicRunCommand(val template: String) : DialogAction

    @JvmRecord
    data class DynamicCustom(val id: String, val additions: Map<String, Any> = emptyMap()) : DialogAction
}
